src/ManagerV2/SharedPersonalDataManager.php

```diff
@@ -0,0 +1,113 @@
+        private readonly MailerService $mailerService,
+        private readonly UserRepository $userRepository,
+        private readonly RequestStack $requestStack,
+        private readonly Security $security,
+    ) {
+    }
+
+    public function generateSharedPersonalDataAndSendEmail(DonneePersonnelle $personalData, string $email, string $locale = 'fr'): void
+    {
+        $user = $this->getUser();
+        if (!$user || !$personalData->getBeneficiaire()) {
+            $this->addFlashMessage('danger', 'user_not_found');
+
+            return;
+        }
+        $sharedPersonalData = $this->factory->generateSharedPersonalData($user, $personalData, $email, $locale);
+        $this->mailerService->sendSharedDocumentLink($sharedPersonalData, $email);
+
+        $this->addFlashMessage('success', 'share_document_success');
+    }
+
+    public function validateTokenAndFetchPersonalData(string $token): ?SharedPersonalData
+    {
+        $decodedToken = $this->decodeSharedPersonalDataToken($token);
+        [
+            'selector' => $selector,
+            'user_id' => $userId,
+            'personal_data_id' => $personalDataId,
+            'personal_data_class' => $personalDataClass,
+        ] = $decodedToken;
+
+        /** @var ?SharedPersonalData $sharedPersonalData * */
+        $sharedPersonalData = $this->getSharedPersonalDataRepository($personalDataClass)?->findOneBy(['selector' => $selector]);
+
+        $errorMessage = $this->getErrorMessage($sharedPersonalData, (int) $userId, (string) $personalDataClass, (int) $personalDataId);
+
+        if ($errorMessage) {
+            $this->addFlashMessage('danger', $errorMessage);
+        }
+
+        return $sharedPersonalData;
+    }
+
+    private function getErrorMessage(?SharedPersonalData $sharedPersonalData, int $userId, string $personalDataClass, int $personalDataId): ?string
+    {
+        return match (true) {
+            !$sharedPersonalData => 'share_document_error_no_shared_document',
+            $sharedPersonalData->isExpired() => 'share_document_error_expired',
+            !$this->userRepository->find($userId) || !$this->getPersonalDataRepository($personalDataClass)?->find($personalDataId) => 'share_document_error_not_found',
+            default => null,
+        };
+    }
+
+    /**
+     * @return string[]
+     */
+    private function decodeSharedPersonalDataToken(string $token): array
+    {
+        return json_decode(base64_decode(str_replace('_', '/', str_replace('-', '+', explode('.', $token)[1]))), true);
+    }
+
+    /**
+     * @return EntityRepository<SharedDocument>|EntityRepository<SharedFolder>|null
+     */
+    private function getSharedPersonalDataRepository(string $personalDataClass): ?EntityRepository
+    {
+        return match ($personalDataClass) {
+            Dossier::class => $this->em->getRepository(SharedFolder::class),
+            Document::class => $this->em->getRepository(SharedDocument::class),
```

Comment:
Tu peux aussi rajouter une méthode `$personalDataClass->getRepositoryClassFQCN()`

Reply:
et du coup ici `return $this->em->getRepository($data->getRepositoryClassFcqn)`

Reply:
C'est ce que j'avais fait au départ, mais on avait des erreurs phpstan malheureusement